### PR TITLE
reorganize searching and removing stuff in MeshLib

### PR DESCRIPTION
**Pull request by norihiro-w**
This PR suggests reorganizing the following tools in MeshLib
- ElementExtraction
- MeshSearcher
- removeMeshNodes
- MeshValidation

and decomposing them into the following new classes
- NodeSearch, ElementSearch
- RemoveMeshComponents

Note: #762 is also included

### CODE REVIEW

**Comment by endJunction** [comment]
Are there any algorithm changes?

**Comment by norihiro-w** [comment]
yes, I changed algorithms in removeNodes() and removeElements() because there are some bugs. New tests (TestRemove.cpp) are added to confirm their behavior.

File: MeshLib/MeshSearch/NodeSearch.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+NodeSearch::NodeSearch(const MeshLib::Mesh &mesh)
+	: _mesh(mesh)
+{
+}
+
+NodeSearch::~NodeSearch()
+{
+}
+
+std::size_t NodeSearch::searchByElementIDs(const std::vector<std::size_t> &elements, bool only_match_all_connected_elements)
```
Comment (norihiro-w) [inline MeshLib/MeshSearch/NodeSearch.cpp]: please review also this function. Second algorithm is added for the case `only_match_all_connected_elements==true`

File: Tests/MeshLib/TestRemove.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/LICENSE.txt
+ */
+
+#include <memory>
+
+#include "gtest/gtest.h"
+
+#include "Mesh.h"
```
Comment (endJunction) [inline Tests/MeshLib/TestRemove.cpp]: Full include paths...

File: MeshLib/MeshSearch/ElementSearch.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ *
+ */
+
+#ifndef ELEMENTSEARCH_H_
+#define ELEMENTSEARCH_H_
+
+#include <limits>
+#include <vector>
+
+#include "GeoLib/AABB.h"
```
Comment (endJunction) [inline MeshLib/MeshSearch/ElementSearch.h]: can be fwd decl.

File: MeshLib/MeshSearch/ElementSearch.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+#include "MeshLib/MeshEnums.h"
+
+namespace MeshLib {
+
+// forward declarations
+class Mesh;
+class Element;
+
+/// Element search class
+class ElementSearch
```
Comment (endJunction) [inline MeshLib/MeshSearch/ElementSearch.h]: final

File: MeshLib/MeshSearch/ElementSearch.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+{
+public:
+	explicit ElementSearch(const MeshLib::Mesh &mesh);
+
+	~ElementSearch();
```
Comment (endJunction) [inline MeshLib/MeshSearch/ElementSearch.h]: dtor is not explicitly needed.

File: MeshLib/MeshSearch/NodeSearch.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	void updateUnion(const std::vector<std::size_t> &vec);
+
+	/// The mesh from which elements should be removed.
+	const MeshLib::Mesh &_mesh;
+	/// The vector of element indices that should be removed.
+	std::vector<std::size_t> _marked_nodes;
+};
+
+/// Create a vector of unique nodes used by given elements.
+std::vector<Node*> getNodes(std::vector<Element*> const& elements);
```
Comment (endJunction) [inline MeshLib/MeshSearch/NodeSearch.h]: better to name it getUniqueNodes.
I also don't like the location of the (quite general) method but do not know a better one. Ideas are welcome.

File: MeshLib/MeshSearch/NodeSearch.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+class Element;
+class Node;
+
+/// Node search class
+class NodeSearch
+{
+public:
+	explicit NodeSearch(const MeshLib::Mesh &mesh);
+
+	~NodeSearch();
```
Comment (endJunction) [inline MeshLib/MeshSearch/NodeSearch.h]: dtor not explicitly needed.

File: MeshLib/MeshSearch/ElementSearch.h (hunk 3 above)
Comment (norihiro-w) [inline MeshLib/MeshSearch/ElementSearch.h]: how can I do it for the template class?

**Comment by norihiro-w** [comment]
It should be noted that this PR does not allow any changes of Mesh object in place and always creates a new Mesh object with modified node and element vectors. This approach is expensive but easier to ensure consistency in a Mesh object (i.e. member variables of Mesh).

**Comment by endJunction** [comment]
I think, now it is a good starting point for further improvements.
:fast_forward:

**Comment by endJunction** [comment]
@norihiro-w needs a rebase after merged ufz/ogs#759 ...

**Pull request merged**